python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+        """
+        self.__data = data
+        self.__data_hash = hash_hashable_map_sequence(data)
+        self.touch()
+
+    @property
+    def identifier(self) -> str:
+        """
+        The identifier/key of the entry
+        """
+        return self.__identifier
+
+    @property
+    def data_hash(self) -> int:
+        """
+        A hash for the data that is contained
+        """
+        return self.__data_hash
+
+    @property
+    def data(self) -> HashableTypeT:
+        """
+        Get the stored data
+        """
+        self.touch()
+        return self.__data
+
+    @data.setter
+    def data(self, new_value: HashableTypeT):
+        self._set_data(data=new_value)
+
+    @property
+    def last_accessed(self) -> datetime:
+        """
+        The last time that this entry was accessed
+        """
+        return self.__last_accessed
+
+    def __eq__(self, other):
+        if isinstance(other, type(self.data)):
+            return self.data == hash_hashable_map_sequence(other)
+
+        if not isinstance(other, CatalogEntry):
+            return False
+
+        return self.data_hash == other.data_hash
+
+    def __hash__(self):
+        return self.data_hash
+
+    def __le__(self, other):
+        if isinstance(other, type(self.data)):
+            return self.data < other or self.data_hash == hash_hashable_map_sequence(other)
+
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} <= {other.__class__.__name__}")
+
+        return self.last_accessed <= other.last_accessed
+
+    def __lt__(self, other):
+        if isinstance(other, type(self.data)):
+            return self.data < other
+
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {type(self.data)} < {type(other)}")
+
+        return self.last_accessed < other.last_accessed
+
+    def __ge__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} >= {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __gt__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} < {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __str__(self):
+        return f"CacheEntry: {self.identifier} @ {self.last_accessed.strftime('%Y-%m-%d %H:%M:%S')}"
+
+    def __repr__(self):
+        return "{" + \
+            f'"identifier": {self.__data}, ' \
+            f'"data_hash": {self.__data_hash}, ' \
+            f'"last_accessed": "{self.last_accessed}"' + \
+            "}"
+
+
+# Part of Issue https://github.com/NOAA-OWP/DMOD/issues/434
+#   "Make metric computations asynchronous by location"
+class InputCatalog(EventfulMap[str, CatalogEntry[HashableTypeT]], typing.Generic[HashableTypeT]):
```

Comment:
Likewise, im more and more convinced that this class can use a `multiprocessing.manager.SharedMemoryManager` to do most of the heavy lifting. I think this class can more or less provide conviences around a `SharedMemoryManager`.

Reply:
I know that this is still a work in progress, so I don't want to weight this down too much. But I think it is super crucial that you mention that this is not thread safe. Really any glaring invariants that need to be true for this to function as expected. For example, who _owns_ the data in the cache? Does the cache own it?